### PR TITLE
DRAFT: Solution to not queue forever dropped events with 2 timeouts

### PR DESCRIPTION
**Pull request by albertogpz**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/ParallelAsyncEventQueueImpl.java [hunk 1]
```diff
@@ -110,6 +110,7 @@ private void start(boolean cleanQueues) {
       logger.info("Started  {}", this);
 
       enqueueTempEvents();
+      startTime = System.currentTimeMillis();
```
Comment (gesterzhou) [inline geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/ParallelAsyncEventQueueImpl.java:113]: startTime should be set after the logger.info

File: geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/ParallelAsyncEventQueueImpl.java [hunk 2]
```diff
@@ -139,6 +140,7 @@ public void stop() {
       system.handleResourceEvent(ResourceEvent.GATEWAYSENDER_STOP, this);
 
       clearTempEventsAfterSenderStopped();
+      stopTime = System.currentTimeMillis();
```
Comment (gesterzhou) [inline geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/ParallelAsyncEventQueueImpl.java:143]: stopTime should be set after log.info()

File: geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/SerialAsyncEventQueueImpl.java [hunk 3]
```diff
@@ -117,6 +117,7 @@ private void start(boolean cleanQueues) {
       logger.info("Started {}", this);
 
       enqueueTempEvents();
+      startTime = System.currentTimeMillis();
```
Comment (gesterzhou) [inline geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/SerialAsyncEventQueueImpl.java:120]: same

File: geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/SerialAsyncEventQueueImpl.java [hunk 4]
```diff
@@ -193,6 +194,7 @@ public void stop() {
     InternalDistributedSystem system =
         (InternalDistributedSystem) this.cache.getDistributedSystem();
     system.handleResourceEvent(ResourceEvent.GATEWAYSENDER_STOP, this);
+    stopTime = System.currentTimeMillis();
```
Comment (gesterzhou) [inline geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/SerialAsyncEventQueueImpl.java:197]: same

File: geode-core/src/main/java/org/apache/geode/internal/cache/xmlcache/ParallelAsyncEventQueueCreation.java [hunk 5]
```diff
@@ -43,13 +43,19 @@ public void distribute(EnumListenerEvent operation, EntryEventImpl event,
       List<Integer> remoteDSIds) {}
 
   @Override
-  public void start() {}
+  public void start() {
```
Comment (gesterzhou) [inline geode-core/src/main/java/org/apache/geode/internal/cache/xmlcache/ParallelAsyncEventQueueCreation.java:46]: we don't need to change the 2 Creation classes.

File: geode-wan/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderImpl.java [hunk 6]
```diff
@@ -102,6 +102,7 @@ private void start(boolean cleanQueues) {
       logger.info("Started {}", this);
 
       enqueueTempEvents();
+      startTime = System.currentTimeMillis();
```
Comment (gesterzhou) [inline geode-wan/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderImpl.java:105]: same

File: geode-wan/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderImpl.java [hunk 7]
```diff
@@ -135,6 +136,7 @@ public void stop() {
       system.handleResourceEvent(ResourceEvent.GATEWAYSENDER_STOP, this);
 
       clearTempEventsAfterSenderStopped();
+      stopTime = System.currentTimeMillis();
```
Comment (gesterzhou) [inline geode-wan/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderImpl.java:139]: same

File: geode-wan/src/main/java/org/apache/geode/internal/cache/wan/serial/SerialGatewaySenderImpl.java [hunk 8]
```diff
@@ -111,6 +111,7 @@ private void start(boolean cleanQueues) {
           .info("Started  {}", this);
 
       enqueueTempEvents();
+      startTime = System.currentTimeMillis();
```
Comment (gesterzhou) [inline geode-wan/src/main/java/org/apache/geode/internal/cache/wan/serial/SerialGatewaySenderImpl.java:114]: same

File: geode-wan/src/main/java/org/apache/geode/internal/cache/wan/serial/SerialGatewaySenderImpl.java [hunk 9]
```diff
@@ -186,6 +187,7 @@ public void stop() {
         (InternalDistributedSystem) this.cache.getDistributedSystem();
     system.handleResourceEvent(ResourceEvent.GATEWAYSENDER_STOP, this);
 
+    stopTime = System.currentTimeMillis();
```
Comment (gesterzhou) [inline geode-wan/src/main/java/org/apache/geode/internal/cache/wan/serial/SerialGatewaySenderImpl.java:190]: same

File: geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANPropagationLoopBackDUnitTest.java [hunk 10]
```diff
@@ -433,6 +455,259 @@ public void unstartedSenderShouldNotAddReceivedEventsIntoTmpDropped() throws Exc
...
+
+    // verify site-ln has not received the events from site-ny because they were dropped
+    vm2.invoke(() -> WANTestBase.validateRegionSize(getTestMethodName() + "_PR", 0));
+    vm4.invoke(() -> WANTestBase.validateRegionSize(getTestMethodName() + "_PR", 0));
+
+    // verify tmpDroppedEvents is 0 now at site-ny
+    vm3.invoke(() -> WANTestBase.verifyTmpDroppedEventSize("ln", 0));
+    vm5.invoke(() -> WANTestBase.verifyTmpDroppedEventSize("ln", 0));
+
   }
```
Comment (gesterzhou) [inline geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANPropagationLoopBackDUnitTest.java:633]: You need to add a test case: 
1) stop sender, after 15 seconds, do a few put(), then retart the sender and make sure these events won't stay in secondary queue for ever.

Comment (albertogpz) [reply geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANPropagationLoopBackDUnitTest.java:633]: That is tested in test case restartedSenderShouldAddReceivedEventsIntoTmpDropped after the 15 seconds sleep.

File: geode-wan/src/main/java/org/apache/geode/internal/cache/wan/serial/SerialGatewaySenderImpl.java (hunk 9 above)
Comment (albertogpz) [reply geode-wan/src/main/java/org/apache/geode/internal/cache/wan/serial/SerialGatewaySenderImpl.java:190]: ok

File: geode-wan/src/main/java/org/apache/geode/internal/cache/wan/serial/SerialGatewaySenderImpl.java (hunk 8 above)
Comment (albertogpz) [reply geode-wan/src/main/java/org/apache/geode/internal/cache/wan/serial/SerialGatewaySenderImpl.java:114]: ok

File: geode-wan/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderImpl.java (hunk 7 above)
Comment (albertogpz) [reply geode-wan/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderImpl.java:139]: ok

File: geode-wan/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderImpl.java (hunk 6 above)
Comment (albertogpz) [reply geode-wan/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderImpl.java:105]: ok

File: geode-core/src/main/java/org/apache/geode/internal/cache/xmlcache/ParallelAsyncEventQueueCreation.java (hunk 5 above)
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/internal/cache/xmlcache/ParallelAsyncEventQueueCreation.java:46]: ok

File: geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/SerialAsyncEventQueueImpl.java (hunk 4 above)
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/SerialAsyncEventQueueImpl.java:197]: ok

File: geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/SerialAsyncEventQueueImpl.java (hunk 3 above)
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/SerialAsyncEventQueueImpl.java:120]: ok

File: geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/ParallelAsyncEventQueueImpl.java (hunk 2 above)
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/ParallelAsyncEventQueueImpl.java:143]: ok

File: geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/ParallelAsyncEventQueueImpl.java (hunk 1 above)
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/asyncqueue/internal/ParallelAsyncEventQueueImpl.java:113]: I would say before the logger.info.

File: geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANPropagationLoopBackDUnitTest.java [hunk 11]
```diff
@@ -433,6 +455,264 @@ public void unstartedSenderShouldNotAddReceivedEventsIntoTmpDropped() throws Exc
...
+    vm3.invoke(() -> WANTestBase.validateRegionSize(getTestMethodName() + "_PR", 100));
+    vm5.invoke(() -> WANTestBase.validateRegionSize(getTestMethodName() + "_PR", 100));
+
+    // verify tmpDroppedEvents is 0 now at site-ln
+    vm2.invoke(() -> WANTestBase.verifyTmpDroppedEventSize("ny", 0));
+    vm4.invoke(() -> WANTestBase.verifyTmpDroppedEventSize("ny", 0));
+
+    // stop sender on site-ny
+    vm2.invoke(() -> stopSender("ny"));
+    vm4.invoke(() -> stopSender("ny"));
```
Comment (jinmeiliao) [inline geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANPropagationLoopBackDUnitTest.java:522]: I thought the test should somehow just stop just the primary gatewaysender, not the secondary. and sleep for 15 minutes, do some puts, and make sure events in secondary sender is all drained....

Comment (albertogpz) [reply geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANPropagationLoopBackDUnitTest.java:522]: If we stop just the primary, wait for 15 seconds and then do the puts, the primary will not longer store the dropped events and when it is started again will not send the events to the secondary to drain them from the queues.
That test case will not work with this solution.
The solution tries to solve the problem when you stop all gateway sender instances together with a very small difference in time among them.

Comment (jinmeiliao) [reply geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANPropagationLoopBackDUnitTest.java:522]: I think that's the test case Gester mentioned in the comment as well.

Comment (albertogpz) [reply geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelWANPropagationLoopBackDUnitTest.java:522]: I understood he meant to stop the complete gateway sender, not just the primary.
It is not possible to solve all the problems if you do not stop in a short amount of time all instances. Either queues will not be drained if the primary is stopped but the secondary keeps receiving events after 15 seconds (with this solution) or the memory of the primary will grow indefinitely storing dropped events while it is not restarted (the latter would happen if we do not apply the fix proposed).

**Comment by onichols-pivotal** [comment]
last activity was some while ago, is this PR still active?

**Comment by albertogpz** [comment]
> last activity was some while ago, is this PR still active?

It is not anymore. The problem has been solved in GEODE-8491.

**Comment by albertogpz** [comment]
The solution in this PR is not valid. The good one can be found in GEODE-8491.

**Pull request closed without merging**